### PR TITLE
Removes unnecessary recalculation

### PR DESCRIPTION
**Pull request by koply**
[contributing]: https://github.com/DV8FromTheWorld/JDA/wiki/5%29-Contributing

## Pull Request Etiquette

- [x] I have checked the PRs for upcoming features/bug fixes.
- [x] I have read the [contributing guidelines][contributing].

### Changes

- [x] Internal code
- [ ] Library interface (affecting end-user code) 
- [ ] Documentation

## Description

Removed recalculation of all listeners on registration at AnnotatedEventManager. Also added the varargs for listener registrar method. A usable edit with no negative effects.

The token cannot be empty. Because pre-checked field. Removed unnecessary check.

### CODE REVIEW

File: src/main/java/net/dv8tion/jda/api/hooks/AnnotatedEventManager.java [hunk 1]
```diff
@@ -55,11 +55,12 @@
     private final Map<Class<?>, Map<Object, List<Method>>> methods = new ConcurrentHashMap<>();
 
     @Override
-    public void register(@Nonnull Object listener)
+    public void register(@Nonnull Object...pListeners)
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/hooks/AnnotatedEventManager.java:58]: This changes the method signature in a breaking way. The ABI method signature becomes an array which means all code referencing this method must be recompiled, as such this is breaking.

Comment (MusaBrt) [reply src/main/java/net/dv8tion/jda/api/hooks/AnnotatedEventManager.java:58]: I'll edit it so that it doesn't change the method signature.

**Comment by MinnDevelopment** [comment]
Should we add unit tests for this?

**Comment by MusaBrt** [comment]
I don't think it should be. My attempts works well. The MessageListenerExample.java works fine.

File: src/main/java/net/dv8tion/jda/internal/JDAImpl.java [hunk 2]
```diff
@@ -278,7 +278,7 @@ public int login(String gatewayUrl, ShardInfo shardInfo, Compression compression
 
         String token = authConfig.getToken();
         setStatus(Status.LOGGING_IN);
-        if (token == null || token.isEmpty())
+        if (token.isEmpty())
```
Comment (Sanduhr32) [inline src/main/java/net/dv8tion/jda/internal/JDAImpl.java:281]: THIS WILL NPE.

**Review by Sanduhr32 — CHANGES_REQUESTED** [review]
I think this chance should be reverted.

File: src/main/java/net/dv8tion/jda/internal/JDAImpl.java (hunk 2 above)
Comment (MusaBrt) [reply src/main/java/net/dv8tion/jda/internal/JDAImpl.java:281]: Check this class. Nothing like that will happen. 

https://github.com/DV8FromTheWorld/JDA/blob/development/src/main/java/net/dv8tion/jda/internal/utils/config/AuthorizationConfig.java

**Comment by MusaBrt** [comment]
Nothing like that will happen.

https://github.com/DV8FromTheWorld/JDA/blob/development/src/main/java/net/dv8tion/jda/internal/utils/config/AuthorizationConfig.java

bump.

**Comment by Sanduhr32** [comment]
I guess, but well

**Comment by DV8FromTheWorld** [comment]
This needs to be rebased against `master`

File: src/main/java/net/dv8tion/jda/api/hooks/AnnotatedEventManager.java [hunk 3]
```diff
@@ -114,31 +114,36 @@ private void updateMethods()
...
         {
-            boolean isClass = listener instanceof Class;
-            Class<?> c = isClass ? (Class) listener : listener.getClass();
-            Method[] allMethods = c.getDeclaredMethods();
-            for (Method m : allMethods)
+            addMethod(listener);
+        }
+    }
+
+    private void addMethod(@Nonnull Object listener)
```
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/api/hooks/AnnotatedEventManager.java:121]: Lets rename this `scanListenerForMethods` or something similar as `addMethod` doesn't really do it for me.

**Comment by MinnDevelopment** [comment]
Superseded by #2454

**Pull request closed without merging**